In [1]:
##setup

import glob
import os
import pandas as pd

processed_dir = '/Users/audriswong/data-portfolio/projects/seattle-checkouts/data/processed'

## determine uniqueness of physical books dataset

### is checkout data unique at monthly grain per title?  ISBN?

In [2]:
#test uniqueness - is each book isbn unique per month?
physical_checkouts = pd.read_parquet(f'{processed_dir}/physical_checkouts.parquet')

# Count how many rows exist per title per month per year
isbn_counts = physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK') &
    (physical_checkouts['checkoutyear'] == 2024) &
    (physical_checkouts['checkoutmonth'] == 12)
].groupby(['isbn', 'checkoutyear', 'checkoutmonth']).size().reset_index(name='row_count')

# If max row_count = 1, titles are unique per month/year
print(f"Max rows per ISBN per month: {isbn_counts['row_count'].max()}")
print(f"ISBNs with duplicates: {(isbn_counts['row_count'] > 1).sum()}")
print(f"Total unique ISBNs: {len(isbn_counts):,}")

Max rows per ISBN per month: 5
ISBNs with duplicates: 22077
Total unique ISBNs: 59,189


In [3]:
#test uniqueness - is each book title unique per month?
physical_checkouts = pd.read_parquet(f'{processed_dir}/physical_checkouts.parquet')

# Count how many rows exist per title per month per year
title_counts = physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK') &
    (physical_checkouts['checkoutyear'] == 2024) &
    (physical_checkouts['checkoutmonth'] == 12)
].groupby(['title', 'checkoutyear', 'checkoutmonth']).size().reset_index(name='row_count')

# If max row_count = 1, titles are unique per month/year
print(f"Max rows per title per month: {title_counts['row_count'].max()}")
print(f"Titles with duplicates: {(title_counts['row_count'] > 1).sum()}")
print(f"Total unique titles: {len(title_counts):,}")

Max rows per title per month: 6
Titles with duplicates: 22459
Total unique titles: 59,286


In [4]:
# Show sample duplicate ISBNs
duplicate_isbns = isbn_counts[isbn_counts['row_count'] > 1]['isbn'].head(5).tolist()

# Show the actual rows for those ISBNs
physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK') &
    (physical_checkouts['checkoutyear'] == 2024) &
    (physical_checkouts['checkoutmonth'] == 12) &
    (physical_checkouts['isbn'].isin(duplicate_isbns))
].sort_values('isbn')

,usageclass,checkouttype,materialtype,checkoutyear,checkoutmonth,checkouts,title,creator,subjects,publisher,publicationyear,isbn,materialtype_upper,title_clean,creator_clean,title_book_clean,work_key,month_date
4450316,Physical,Horizon,BOOK,2024,12,10,Bog Myrtle / Sid Sharp.,"Sharp, Sid","Siblings Comic books strips etc, Spiders Comic...","Annick Press,",2024.,", 1773218921, 1773218948, 9781773218922, 97817...",BOOK,BOG MYRTLE / SID SHARP.,"SHARP, SID",BOG MYRTLE,"BOG MYRTLE / SID SHARP.||SHARP, SID",2024-12-01
8784228,Physical,Horizon,BOOK,2024,12,10,Bog Myrtle / Sid Sharp.,"Sharp, Sid","Siblings Comic books strips etc, Spiders Comic...","Annick Press,",2024.,", 1773218921, 1773218948, 9781773218922, 97817...",BOOK,BOG MYRTLE / SID SHARP.,"SHARP, SID",BOG MYRTLE,"BOG MYRTLE / SID SHARP.||SHARP, SID",2024-12-01
161357,Physical,Horizon,BOOK,2024,12,2,Spring story / Jill Barklem.,"Barklem, Jill","Mice Juvenile fiction, Birthdays Juvenile fict...","HarperCollins Children's Books,",2020.,"0001839225, 0008241171, 9780001839229, 9780008...",BOOK,SPRING STORY / JILL BARKLEM.,"BARKLEM, JILL",SPRING STORY,"SPRING STORY / JILL BARKLEM.||BARKLEM, JILL",2024-12-01
3038426,Physical,Horizon,BOOK,2024,12,2,Spring story / Jill Barklem.,"Barklem, Jill","Mice Juvenile fiction, Birthdays Juvenile fict...","HarperCollins Children's Books,",2020.,"0001839225, 0008241171, 9780001839229, 9780008...",BOOK,SPRING STORY / JILL BARKLEM.,"BARKLEM, JILL",SPRING STORY,"SPRING STORY / JILL BARKLEM.||BARKLEM, JILL",2024-12-01
6983022,Physical,Horizon,BOOK,2024,12,1,Junior : Griffey on Griffey / written by Ken G...,"Griffey, Ken, Jr.","Griffey Ken Jr, Seattle Mariners Baseball team...","CollinsPublishers ,",[1997],"0002252198, 0006491545",BOOK,JUNIOR : GRIFFEY ON GRIFFEY / WRITTEN BY KEN G...,"GRIFFEY, KEN, JR.",JUNIOR : GRIFFEY ON GRIFFEY,JUNIOR : GRIFFEY ON GRIFFEY / WRITTEN BY KEN G...,2024-12-01
7133258,Physical,Horizon,BOOK,2024,12,1,Junior : Griffey on Griffey / written by Ken G...,"Griffey, Ken, Jr.","Griffey Ken Jr, Seattle Mariners Baseball team...","CollinsPublishers ,",[1997],"0002252198, 0006491545",BOOK,JUNIOR : GRIFFEY ON GRIFFEY / WRITTEN BY KEN G...,"GRIFFEY, KEN, JR.",JUNIOR : GRIFFEY ON GRIFFEY,JUNIOR : GRIFFEY ON GRIFFEY / WRITTEN BY KEN G...,2024-12-01
12001365,Physical,Horizon,BOOK,2024,12,1,Junior : Griffey on Griffey / written by Ken G...,"Griffey, Ken, Jr.","Griffey Ken Jr, Seattle Mariners Baseball team...","CollinsPublishers ,",[1997],"0002252198, 0006491545",BOOK,JUNIOR : GRIFFEY ON GRIFFEY / WRITTEN BY KEN G...,"GRIFFEY, KEN, JR.",JUNIOR : GRIFFEY ON GRIFFEY,JUNIOR : GRIFFEY ON GRIFFEY / WRITTEN BY KEN G...,2024-12-01
10640948,Physical,Horizon,BOOK,2024,12,8,Mog the forgetful cat / written and illustrate...,"Kerr, Judith",Mog Fictitious character from Kerr Juvenile fi...,"HarperCollins Children's Books,",2013.,"000717134X, 0007211341, 0007526326, 9780007171...",BOOK,MOG THE FORGETFUL CAT / WRITTEN AND ILLUSTRATE...,"KERR, JUDITH",MOG THE FORGETFUL CAT,MOG THE FORGETFUL CAT / WRITTEN AND ILLUSTRATE...,2024-12-01
12138742,Physical,Horizon,BOOK,2024,12,8,Mog the forgetful cat / written and illustrate...,"Kerr, Judith",Mog Fictitious character from Kerr Juvenile fi...,"HarperCollins Children's Books,",2013.,"000717134X, 0007211341, 0007526326, 9780007171...",BOOK,MOG THE FORGETFUL CAT / WRITTEN AND ILLUSTRATE...,"KERR, JUDITH",MOG THE FORGETFUL CAT,MOG THE FORGETFUL CAT / WRITTEN AND ILLUSTRATE...,2024-12-01
5274424,Physical,Horizon,BOOK,2024,12,3,Night of power : the betrayal of the Middle Ea...,"Fisk, Robert","Middle East History 21st century, Middle East ...","4th Estate,",2024.,"0007255489, 0007350619, 9780007255481, 9780007...",BOOK,NIGHT OF POWER : THE BETRAYAL OF THE MIDDLE EA...,"FISK, ROBERT",NIGHT OF POWER : THE BETRAYAL OF THE MIDDLE EAST,NIGHT OF POWER : THE BETRAYAL OF THE MIDDLE EA...,2024-12-01


In [5]:
# Show sample duplicate titles
duplicate_titles = title_counts[title_counts['row_count'] > 1]['title'].head(5).tolist()

# Show the actual rows for those titles
physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK') &
    (physical_checkouts['checkoutyear'] == 2024) &
    (physical_checkouts['checkoutmonth'] == 12) &
    (physical_checkouts['title'].isin(duplicate_titles))
].sort_values('title')

,usageclass,checkouttype,materialtype,checkoutyear,checkoutmonth,checkouts,title,creator,subjects,publisher,publicationyear,isbn,materialtype_upper,title_clean,creator_clean,title_book_clean,work_key,month_date
1044909,Physical,Horizon,BOOK,2024,12,2,!Vamos a celebrar! : fechas especiales en el m...,"DePalma, Kate","Holidays Juvenile literature, Festivals Juveni...","Barefoot Books,",2024.,9798888592090,BOOK,!VAMOS A CELEBRAR! : FECHAS ESPECIALES EN EL M...,"DEPALMA, KATE",!VAMOS A CELEBRAR! : FECHAS ESPECIALES EN EL M...,!VAMOS A CELEBRAR! : FECHAS ESPECIALES EN EL M...,2024-12-01
1619132,Physical,Horizon,BOOK,2024,12,2,!Vamos a celebrar! : fechas especiales en el m...,"DePalma, Kate","Holidays Juvenile literature, Festivals Juveni...","Barefoot Books,",2024.,9798888592090,BOOK,!VAMOS A CELEBRAR! : FECHAS ESPECIALES EN EL M...,"DEPALMA, KATE",!VAMOS A CELEBRAR! : FECHAS ESPECIALES EN EL M...,!VAMOS A CELEBRAR! : FECHAS ESPECIALES EN EL M...,2024-12-01
340564,Physical,Horizon,BOOK,2024,12,2,"""All the real Indians died off"" : and 20 other...","Dunbar-Ortiz, Roxanne, 1938-",Indians of North America Historiography,"Beacon Press,",2016.,"0807062650, 9780807062654",BOOK,"""ALL THE REAL INDIANS DIED OFF"" : AND 20 OTHER...","DUNBAR-ORTIZ, ROXANNE","""ALL THE REAL INDIANS DIED OFF"" : AND 20 OTHER...","""ALL THE REAL INDIANS DIED OFF"" : AND 20 OTHER...",2024-12-01
5271809,Physical,Horizon,BOOK,2024,12,2,"""All the real Indians died off"" : and 20 other...","Dunbar-Ortiz, Roxanne, 1938-",Indians of North America Historiography,"Beacon Press,",2016.,"0807062650, 9780807062654",BOOK,"""ALL THE REAL INDIANS DIED OFF"" : AND 20 OTHER...","DUNBAR-ORTIZ, ROXANNE","""ALL THE REAL INDIANS DIED OFF"" : AND 20 OTHER...","""ALL THE REAL INDIANS DIED OFF"" : AND 20 OTHER...",2024-12-01
1045180,Physical,Horizon,BOOK,2024,12,1,"""Better living"" : advertising, media and the n...","Bird, William L., Jr.","Advertising United States History, Public rela...","Northwestern University Press,",1999.,"0810115859, 0810115867",BOOK,"""BETTER LIVING"" : ADVERTISING, MEDIA AND THE N...","BIRD, WILLIAM L., JR.","""BETTER LIVING"" : ADVERTISING, MEDIA AND THE N...","""BETTER LIVING"" : ADVERTISING, MEDIA AND THE N...",2024-12-01
1619403,Physical,Horizon,BOOK,2024,12,1,"""Better living"" : advertising, media and the n...","Bird, William L., Jr.","Advertising United States History, Public rela...","Northwestern University Press,",1999.,"0810115859, 0810115867",BOOK,"""BETTER LIVING"" : ADVERTISING, MEDIA AND THE N...","BIRD, WILLIAM L., JR.","""BETTER LIVING"" : ADVERTISING, MEDIA AND THE N...","""BETTER LIVING"" : ADVERTISING, MEDIA AND THE N...",2024-12-01
341642,Physical,Horizon,BOOK,2024,12,1,"""D"" is for deadbeat / Sue Grafton.","Grafton, Sue",Millhone Kinsey Fictitious character 1950 Fict...,"H. Holt,",[1987],"0805002480, 9780805002485",BOOK,"""D"" IS FOR DEADBEAT / SUE GRAFTON.","GRAFTON, SUE","""D"" IS FOR DEADBEAT","""D"" IS FOR DEADBEAT / SUE GRAFTON.||GRAFTON, SUE",2024-12-01
5272887,Physical,Horizon,BOOK,2024,12,1,"""D"" is for deadbeat / Sue Grafton.","Grafton, Sue",Millhone Kinsey Fictitious character 1950 Fict...,"H. Holt,",[1987],"0805002480, 9780805002485",BOOK,"""D"" IS FOR DEADBEAT / SUE GRAFTON.","GRAFTON, SUE","""D"" IS FOR DEADBEAT","""D"" IS FOR DEADBEAT / SUE GRAFTON.||GRAFTON, SUE",2024-12-01
1532500,Physical,Horizon,BOOK,2024,12,2,"""I'm not Santa!"" / Jonathan Allen.","Allen, Jonathan, 1957-","Mistaken identity Juvenile fiction, Santa Clau...","Hyperion Books for Children,",2008.,"1423113004, 9781423113003",BOOK,"""I'M NOT SANTA!"" / JONATHAN ALLEN.","ALLEN, JONATHAN","""I'M NOT SANTA!""","""I'M NOT SANTA!"" / JONATHAN ALLEN.||ALLEN, JON...",2024-12-01
3302874,Physical,Horizon,BOOK,2024,12,2,"""I'm not Santa!"" / Jonathan Allen.","Allen, Jonathan, 1957-","Mistaken identity Juvenile fiction, Santa Clau...","Hyperion Books for Children,",2008.,"1423113004, 9781423113003",BOOK,"""I'M NOT SANTA!"" / JONATHAN ALLEN.","ALLEN, JONATHAN","""

In [6]:
# Find titles that appear multiple times but have different ISBNs
isbn_check = physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK') &
    (physical_checkouts['checkoutyear'] == 2024) &
    (physical_checkouts['checkoutmonth'] == 12)
].groupby('title').agg(
    row_count=('title', 'count'),
    unique_isbns=('isbn', 'nunique')
).reset_index()

# Duplicate titles with different ISBNs
different_isbns = isbn_check[
    (isbn_check['row_count'] > 1) &
    (isbn_check['unique_isbns'] > 1)
]

print(f"Titles with duplicates and different ISBNs: {len(different_isbns):,}")

# Show sample rows from source data for these titles
sample_titles = different_isbns['title'].head(3).tolist()
physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK') &
    (physical_checkouts['checkoutyear'] == 2024) &
    (physical_checkouts['checkoutmonth'] == 12) &
    (physical_checkouts['title'].isin(sample_titles))
].sort_values('title')

Titles with duplicates and different ISBNs: 560


,usageclass,checkouttype,materialtype,checkoutyear,checkoutmonth,checkouts,title,creator,subjects,publisher,publicationyear,isbn,materialtype_upper,title_clean,creator_clean,title_book_clean,work_key,month_date
8177376,Physical,Horizon,BOOK,2024,12,1,'Twas the bite before Christmas / David Rosenf...,"Rosenfelt, David","Carpenter Andy Fictitious character Fiction, D...","Thorndike Press, a part of Gale, a Cengage Com...",2023.,9798885794053,BOOK,'TWAS THE BITE BEFORE CHRISTMAS / DAVID ROSENF...,"ROSENFELT, DAVID",'TWAS THE BITE BEFORE CHRISTMAS,'TWAS THE BITE BEFORE CHRISTMAS / DAVID ROSENF...,2024-12-01
9571727,Physical,Horizon,BOOK,2024,12,4,'Twas the bite before Christmas / David Rosenf...,"Rosenfelt, David","Carpenter Andy Fictitious character Fiction, D...","Minotaur Books,",2023.,"1250828848, 9781250828842",BOOK,'TWAS THE BITE BEFORE CHRISTMAS / DAVID ROSENF...,"ROSENFELT, DAVID",'TWAS THE BITE BEFORE CHRISTMAS,'TWAS THE BITE BEFORE CHRISTMAS / DAVID ROSENF...,2024-12-01
11788203,Physical,Horizon,BOOK,2024,12,4,'Twas the bite before Christmas / David Rosenf...,"Rosenfelt, David","Carpenter Andy Fictitious character Fiction, D...","Minotaur Books,",2023.,"1250828848, 9781250828842",BOOK,'TWAS THE BITE BEFORE CHRISTMAS / DAVID ROSENF...,"ROSENFELT, DAVID",'TWAS THE BITE BEFORE CHRISTMAS,'TWAS THE BITE BEFORE CHRISTMAS / DAVID ROSENF...,2024-12-01
162212,Physical,Horizon,BOOK,2024,12,1,1st case / James Patterson and Chris Tebbetts.,"Patterson, James, 1947-",United States Federal Bureau of Investigation ...,"Little, Brown and Company,",2020.,"0316418188, 9780316418188",BOOK,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS.,"PATTERSON, JAMES",1ST CASE,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS....,2024-12-01
3039281,Physical,Horizon,BOOK,2024,12,1,1st case / James Patterson and Chris Tebbetts.,"Patterson, James, 1947-",United States Federal Bureau of Investigation ...,"Little, Brown and Company,",2020.,"0316418188, 9780316418188",BOOK,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS.,"PATTERSON, JAMES",1ST CASE,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS....,2024-12-01
6984291,Physical,Horizon,BOOK,2024,12,1,1st case / James Patterson and Chris Tebbetts.,"Patterson, James, 1947-",United States Federal Bureau of Investigation ...,"Little, Brown and Company,",2020.,"0316418315, 1648380360, 9780316418317, 9781648...",BOOK,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS.,"PATTERSON, JAMES",1ST CASE,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS....,2024-12-01
7134527,Physical,Horizon,BOOK,2024,12,1,1st case / James Patterson and Chris Tebbetts.,"Patterson, James, 1947-",United States Federal Bureau of Investigation ...,"Little, Brown and Company,",2020.,"0316418315, 1648380360, 9780316418317, 9781648...",BOOK,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS.,"PATTERSON, JAMES",1ST CASE,1ST CASE / JAMES PATTERSON AND CHRIS TEBBETTS....,2024-12-01
1046099,Physical,Horizon,BOOK,2024,12,3,2054 : a novel / Elliot Ackerman and Admiral J...,"Ackerman, Elliot",United States Politics and government 2021 Fic...,"Penguin Press,",2024.,"0593489861, 0593830857, 9780593489864, 9780593...",BOOK,2054 : A NOVEL / ELLIOT ACKERMAN AND ADMIRAL J...,"ACKERMAN, ELLIOT",2054 : A NOVEL,2054 : A NOVEL / ELLIOT ACKERMAN AND ADMIRAL J...,2024-12-01
1047469,Physical,Horizon,BOOK,2024,12,1,2054 : a novel / Elliot Ackerman and Admiral J...,"Ackerman, Elliot","Large type books, United States Politics and g...","Thorndike Press, a part of Gale, a Cengage Com...",2024.,"1420516213, 9781420516210",BOOK,2054 : A NOVEL / ELLIOT ACKERMAN AND ADMIRAL J...,"ACKERMAN, ELLIOT",2054 : A NOVEL,2054 : A NOVEL / ELLIOT ACKERMAN AND ADMIRAL J...,2024-12-01
1620322,Physical,Horizon,BOOK,2024,12,3,2054 : a novel / Elliot Ackerman and Admiral J...,"Ackerman, Elliot",United States Politics and government 2021 Fic...,"Penguin Press,",2024.,"0593489861, 0593830857, 9780593489864, 9780593...",BOOK,2054 : A NOVEL / ELLIOT ACKERMAN AND ADMIRAL J...,"ACKERMAN, ELLIOT",2054 : A NOVEL,2054 : A 

## do the same checks for digital ebooks

In [7]:
#test uniqueness - is each ebook isbn unique per month?
digital_checkouts = pd.read_parquet(f'{processed_dir}/digital_checkouts.parquet')

# Count how many rows exist per title per month per year
digital_isbn_counts = digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK') &
    (digital_checkouts['checkoutyear'] == 2024) &
    (digital_checkouts['checkoutmonth'] == 12)
].groupby(['isbn', 'checkoutyear', 'checkoutmonth']).size().reset_index(name='row_count')

# If max row_count = 1, titles are unique per month/year
print(f"Max rows per ISBN per month: {digital_isbn_counts['row_count'].max()}")
print(f"ISBNs with duplicates: {(digital_isbn_counts['row_count'] > 1).sum()}")
print(f"Total unique ISBNs: {len(digital_isbn_counts):,}")

Max rows per ISBN per month: 4
ISBNs with duplicates: 21133
Total unique ISBNs: 56,791


In [8]:
#test uniqueness - is each ebook title unique per month?
digital_checkouts = pd.read_parquet(f'{processed_dir}/digital_checkouts.parquet')

# Count how many rows exist per title per month per year
digital_title_counts = digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK') &
    (digital_checkouts['checkoutyear'] == 2024) &
    (digital_checkouts['checkoutmonth'] == 12)
].groupby(['title', 'checkoutyear', 'checkoutmonth']).size().reset_index(name='row_count')

# If max row_count = 1, titles are unique per month/year
print(f"Max rows per ISBN per month: {digital_title_counts['row_count'].max()}")
print(f"ISBNs with duplicates: {(digital_title_counts['row_count'] > 1).sum()}")
print(f"Total unique ISBNs: {len(digital_title_counts):,}")

Max rows per ISBN per month: 11
ISBNs with duplicates: 21498
Total unique ISBNs: 56,071


In [9]:
# Show sample duplicate ISBNs
duplicate_digital_isbns = digital_isbn_counts[digital_isbn_counts['row_count'] > 1]['isbn'].head(5).tolist()

# Show the actual rows for those ISBNs
digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK') &
    (digital_checkouts['checkoutyear'] == 2024) &
    (digital_checkouts['checkoutmonth'] == 12) &
    (digital_checkouts['isbn'].isin(duplicate_digital_isbns))
].sort_values('isbn')

,usageclass,checkouttype,materialtype,checkoutyear,checkoutmonth,checkouts,title,creator,subjects,publisher,publicationyear,isbn,materialtype_upper,title_clean,creator_clean,title_book_clean,work_key,month_date
4022438,Digital,OverDrive,EBOOK,2024,12,1,Pride and Prejudice,Jane Austen,"Classic Literature, Fiction",Barnes & Noble World Digital Library,2005,0594081610,EBOOK,PRIDE AND PREJUDICE,JANE AUSTEN,PRIDE AND PREJUDICE,PRIDE AND PREJUDICE||JANE AUSTEN,2024-12-01
7634602,Digital,OverDrive,EBOOK,2024,12,1,Pride and Prejudice,Jane Austen,"Classic Literature, Fiction",Barnes & Noble World Digital Library,2005,0594081610,EBOOK,PRIDE AND PREJUDICE,JANE AUSTEN,PRIDE AND PREJUDICE,PRIDE AND PREJUDICE||JANE AUSTEN,2024-12-01
200805,Digital,OverDrive,EBOOK,2024,12,3,Sense and Sensibility (World Digital Library E...,Jane Austen,"Classic Literature, Fiction",Barnes & Noble World Digital Library,2005,0594082714,EBOOK,SENSE AND SENSIBILITY (WORLD DIGITAL LIBRARY E...,JANE AUSTEN,SENSE AND SENSIBILITY (WORLD DIGITAL LIBRARY E...,SENSE AND SENSIBILITY (WORLD DIGITAL LIBRARY E...,2024-12-01
2805095,Digital,OverDrive,EBOOK,2024,12,3,Sense and Sensibility (World Digital Library E...,Jane Austen,"Classic Literature, Fiction",Barnes & Noble World Digital Library,2005,0594082714,EBOOK,SENSE AND SENSIBILITY (WORLD DIGITAL LIBRARY E...,JANE AUSTEN,SENSE AND SENSIBILITY (WORLD DIGITAL LIBRARY E...,SENSE AND SENSIBILITY (WORLD DIGITAL LIBRARY E...,2024-12-01
3001758,Digital,OverDrive,EBOOK,2024,12,1,Oliver Twist (World Digital Library Edition),Charles Dickens,"Fiction, Literature",Barnes & Noble World Digital Library,2005,059408301X,EBOOK,OLIVER TWIST (WORLD DIGITAL LIBRARY EDITION),CHARLES DICKENS,OLIVER TWIST (WORLD DIGITAL LIBRARY EDITION),OLIVER TWIST (WORLD DIGITAL LIBRARY EDITION)||...,2024-12-01
6719630,Digital,OverDrive,EBOOK,2024,12,1,Oliver Twist (World Digital Library Edition),Charles Dickens,"Fiction, Literature",Barnes & Noble World Digital Library,2005,059408301X,EBOOK,OLIVER TWIST (WORLD DIGITAL LIBRARY EDITION),CHARLES DICKENS,OLIVER TWIST (WORLD DIGITAL LIBRARY EDITION),OLIVER TWIST (WORLD DIGITAL LIBRARY EDITION)||...,2024-12-01
4695366,Digital,OverDrive,EBOOK,2024,12,2,A Little Princess (World Digital Library Edition),Frances Hodgson Burnett,"Juvenile Fiction, Juvenile Literature",Barnes & Noble World Digital Library,2005,0594083567,EBOOK,A LITTLE PRINCESS (WORLD DIGITAL LIBRARY EDITION),FRANCES HODGSON BURNETT,A LITTLE PRINCESS (WORLD DIGITAL LIBRARY EDITION),A LITTLE PRINCESS (WORLD DIGITAL LIBRARY EDITI...,2024-12-01
9860559,Digital,OverDrive,EBOOK,2024,12,2,A Little Princess (World Digital Library Edition),Frances Hodgson Burnett,"Juvenile Fiction, Juvenile Literature",Barnes & Noble World Digital Library,2005,0594083567,EBOOK,A LITTLE PRINCESS (WORLD DIGITAL LIBRARY EDITION),FRANCES HODGSON BURNETT,A LITTLE PRINCESS (WORLD DIGITAL LIBRARY EDITION),A LITTLE PRINCESS (WORLD DIGITAL LIBRARY EDITI...,2024-12-01
3004533,Digital,OverDrive,EBOOK,2024,12,4,A Christmas Carol (World Digital Library Edition),Charles Dickens,"Fiction, Literature",Barnes & Noble World Digital Library,2005,0594083710,EBOOK,A CHRISTMAS CAROL (WORLD DIGITAL LIBRARY EDITION),CHARLES DICKENS,A CHRISTMAS CAROL (WORLD DIGITAL LIBRARY EDITION),A CHRISTMAS CAROL (WORLD DIGITAL LIBRARY EDITI...,2024-12-01
6722405,Digital,OverDrive,EBOOK,2024,12,4,A Christmas Carol (World Digital Library Edition),Charles Dickens,"Fiction, Literature",Barnes & Noble World Digital Library,2005,0594083710,EBOOK,A CHRISTMAS CAROL (WORLD DIGITAL LIBRARY EDITION),CHARLES DICKENS,A CHRISTMAS CAROL (WORLD DIGITAL LIBRARY EDITION),A CHRISTMAS CAROL (WORLD DIGITAL LIBRARY EDITI...,2024-12-01


In [10]:
# Show sample duplicate titles
duplicate_digital_titles = digital_title_counts[digital_title_counts['row_count'] > 1]['title'].head(5).tolist()

# Show the actual rows for those ISBNs
digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK') &
    (digital_checkouts['checkoutyear'] == 2024) &
    (digital_checkouts['checkoutmonth'] == 12) &
    (digital_checkouts['title'].isin(duplicate_digital_titles))
].sort_values('title')

,usageclass,checkouttype,materialtype,checkoutyear,checkoutmonth,checkouts,title,creator,subjects,publisher,publicationyear,isbn,materialtype_upper,title_clean,creator_clean,title_book_clean,work_key,month_date
4024821,Digital,OverDrive,EBOOK,2024,12,1,"""B"" Is for Betsy",Carolyn Haywood,"Juvenile Fiction, Juvenile Literature",HarperCollins Publishers Inc.,2011,9780547671338,EBOOK,"""B"" IS FOR BETSY",CAROLYN HAYWOOD,"""B"" IS FOR BETSY","""B"" IS FOR BETSY||CAROLYN HAYWOOD",2024-12-01
10063161,Digital,OverDrive,EBOOK,2024,12,1,"""B"" Is for Betsy",Carolyn Haywood,"Juvenile Fiction, Juvenile Literature",HarperCollins Publishers Inc.,2011,9780547671338,EBOOK,"""B"" IS FOR BETSY",CAROLYN HAYWOOD,"""B"" IS FOR BETSY","""B"" IS FOR BETSY||CAROLYN HAYWOOD",2024-12-01
9210941,Digital,OverDrive,EBOOK,2024,12,1,"""I Have Nothing to Hide"": And 20 Other Myths A...",Heidi Boghosian,"Politics, Sociology, Nonfiction, Media Studies","Random House, Inc.",2021,9780807061275,EBOOK,"""I HAVE NOTHING TO HIDE"": AND 20 OTHER MYTHS A...",HEIDI BOGHOSIAN,"""I HAVE NOTHING TO HIDE"": AND 20 OTHER MYTHS A...","""I HAVE NOTHING TO HIDE"": AND 20 OTHER MYTHS A...",2024-12-01
10733476,Digital,OverDrive,EBOOK,2024,12,1,"""I Have Nothing to Hide"": And 20 Other Myths A...",Heidi Boghosian,"Politics, Sociology, Nonfiction, Media Studies","Random House, Inc.",2021,9780807061275,EBOOK,"""I HAVE NOTHING TO HIDE"": AND 20 OTHER MYTHS A...",HEIDI BOGHOSIAN,"""I HAVE NOTHING TO HIDE"": AND 20 OTHER MYTHS A...","""I HAVE NOTHING TO HIDE"": AND 20 OTHER MYTHS A...",2024-12-01
4694838,Digital,OverDrive,EBOOK,2024,12,1,"""You Can Tell Just by Looking"": And 20 Other M...",Michael Bronski,"LGBTQIA+ (Nonfiction), Sociology, Nonfiction","Random House, Inc.",2022,9780807042465,EBOOK,"""YOU CAN TELL JUST BY LOOKING"": AND 20 OTHER M...",MICHAEL BRONSKI,"""YOU CAN TELL JUST BY LOOKING"": AND 20 OTHER M...","""YOU CAN TELL JUST BY LOOKING"": AND 20 OTHER M...",2024-12-01
9860031,Digital,OverDrive,EBOOK,2024,12,1,"""You Can Tell Just by Looking"": And 20 Other M...",Michael Bronski,"LGBTQIA+ (Nonfiction), Sociology, Nonfiction","Random House, Inc.",2022,9780807042465,EBOOK,"""YOU CAN TELL JUST BY LOOKING"": AND 20 OTHER M...",MICHAEL BRONSKI,"""YOU CAN TELL JUST BY LOOKING"": AND 20 OTHER M...","""YOU CAN TELL JUST BY LOOKING"": AND 20 OTHER M...",2024-12-01
4019869,Digital,OverDrive,EBOOK,2024,12,4,"""You Just Need to Lose Weight"": And 19 Other M...",Aubrey Gordon,"Self-Improvement, Sociology, Nonfiction","Random House, Inc.",2023,9780807006481,EBOOK,"""YOU JUST NEED TO LOSE WEIGHT"": AND 19 OTHER M...",AUBREY GORDON,"""YOU JUST NEED TO LOSE WEIGHT"": AND 19 OTHER M...","""YOU JUST NEED TO LOSE WEIGHT"": AND 19 OTHER M...",2024-12-01
7632033,Digital,OverDrive,EBOOK,2024,12,4,"""You Just Need to Lose Weight"": And 19 Other M...",Aubrey Gordon,"Self-Improvement, Sociology, Nonfiction","Random House, Inc.",2023,9780807006481,EBOOK,"""YOU JUST NEED TO LOSE WEIGHT"": AND 19 OTHER M...",AUBREY GORDON,"""YOU JUST NEED TO LOSE WEIGHT"": AND 19 OTHER M...","""YOU JUST NEED TO LOSE WEIGHT"": AND 19 OTHER M...",2024-12-01
572638,Digital,OverDrive,EBOOK,2024,12,3,#FashionVictim: A Novel,Amina Akhtar,"Fiction, Suspense, Thriller","Random House, Inc.",2018,9781683318354,EBOOK,#FASHIONVICTIM: A NOVEL,AMINA AKHTAR,#FASHIONVICTIM: A NOVEL,#FASHIONVICTIM: A NOVEL||AMINA AKHTAR,2024-12-01
9860951,Digital,OverDrive,EBOOK,2024,12,3,#FashionVictim: A Novel,Amina Akhtar,"Fiction, Suspense, Thriller","Random House, Inc.",2018,9781683318354,EBOOK,#FASHIONVICTIM: A NOVEL,AMINA AKHTAR,#FASHIONVICTIM: A NOVEL,#FASHIONVICTIM: A NOVEL||AMINA AKHTAR,2024-12-01


In [11]:
# Find titles that appear multiple times but have different ISBNs
digital_isbn_check = digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK') &
    (digital_checkouts['checkoutyear'] == 2024) &
    (digital_checkouts['checkoutmonth'] == 12)
].groupby('title').agg(
    row_count=('title', 'count'),
    unique_isbns=('isbn', 'nunique')
).reset_index()

# Duplicate titles with different ISBNs
different_digital_isbns = digital_isbn_check[
    (digital_isbn_check['row_count'] > 1) &
    (digital_isbn_check['unique_isbns'] > 1)
]

print(f"Titles with duplicates and different ISBNs: {len(different_digital_isbns):,}")

# Show sample rows from source data for these titles
sample_digital_titles = different_digital_isbns['title'].head(3).tolist()
digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK') &
    (digital_checkouts['checkoutyear'] == 2024) &
    (digital_checkouts['checkoutmonth'] == 12) &
    (digital_checkouts['title'].isin(sample_digital_titles))
].sort_values('title')

Titles with duplicates and different ISBNs: 972


,usageclass,checkouttype,materialtype,checkoutyear,checkoutmonth,checkouts,title,creator,subjects,publisher,publicationyear,isbn,materialtype_upper,title_clean,creator_clean,title_book_clean,work_key,month_date
201622,Digital,OverDrive,EBOOK,2024,12,2,'Twas the Night Before Christmas,Clement Clarke Moore,"Classic Literature, Juvenile Fiction, Juvenile...",HarperCollins Publishers Inc.,2011,9780547531175,EBOOK,'TWAS THE NIGHT BEFORE CHRISTMAS,CLEMENT CLARKE MOORE,'TWAS THE NIGHT BEFORE CHRISTMAS,'TWAS THE NIGHT BEFORE CHRISTMAS||CLEMENT CLAR...,2024-12-01
2805912,Digital,OverDrive,EBOOK,2024,12,2,'Twas the Night Before Christmas,Clement Clarke Moore,"Classic Literature, Juvenile Fiction, Juvenile...",HarperCollins Publishers Inc.,2011,9780547531175,EBOOK,'TWAS THE NIGHT BEFORE CHRISTMAS,CLEMENT CLARKE MOORE,'TWAS THE NIGHT BEFORE CHRISTMAS,'TWAS THE NIGHT BEFORE CHRISTMAS||CLEMENT CLAR...,2024-12-01
10064247,Digital,OverDrive,EBOOK,2024,12,3,'Twas the Night Before Christmas,Jane Chapman,"Juvenile Fiction, Juvenile Literature",Sourcebooks,2024,9781464222788,EBOOK,'TWAS THE NIGHT BEFORE CHRISTMAS,JANE CHAPMAN,'TWAS THE NIGHT BEFORE CHRISTMAS,'TWAS THE NIGHT BEFORE CHRISTMAS||JANE CHAPMAN,2024-12-01
366784,Digital,OverDrive,EBOOK,2024,12,3,1919,Eve L. Ewing,"Fiction, Poetry",Lightning Source Inc,2019,9781608466009,EBOOK,1919,EVE L. EWING,1919,1919||EVE L. EWING,2024-12-01
572554,Digital,OverDrive,EBOOK,2024,12,2,1919,John Dos Passos,"Fiction, Literature, Historical Fiction",Open Road Integrated Media,2016,9780544341456,EBOOK,1919,JOHN DOS PASSOS,1919,1919||JOHN DOS PASSOS,2024-12-01
4694180,Digital,OverDrive,EBOOK,2024,12,3,1919,Eve L. Ewing,"Fiction, Poetry",Lightning Source Inc,2019,9781608466009,EBOOK,1919,EVE L. EWING,1919,1919||EVE L. EWING,2024-12-01
9860867,Digital,OverDrive,EBOOK,2024,12,2,1919,John Dos Passos,"Fiction, Literature, Historical Fiction",Open Road Integrated Media,2016,9780544341456,EBOOK,1919,JOHN DOS PASSOS,1919,1919||JOHN DOS PASSOS,2024-12-01
199964,Digital,OverDrive,EBOOK,2024,12,68,1984,George Orwell,"Classic Literature, Fiction, Literature",HarperCollins Publishers Inc.,2011,9780547249643,EBOOK,1984,GEORGE ORWELL,1984,1984||GEORGE ORWELL,2024-12-01
364431,Digital,OverDrive,EBOOK,2024,12,1,1984,George Orwell,"Classic Literature, Fiction",Ink It Digital,2020,9789590308635,EBOOK,1984,GEORGE ORWELL,1984,1984||GEORGE ORWELL,2024-12-01
2804254,Digital,OverDrive,EBOOK,2024,12,68,1984,George Orwell,"Classic Literature, Fiction, Literature",HarperCollins Publishers Inc.,2011,9780547249643,EBOOK,1984,GEORGE ORWELL,1984,1984||GEORGE ORWELL,2024-12-01


### for the sake of this analysis, I'm interested in tracking checkouts per title<>creator, not per title edition (e.g. "Ramona the Brave" 1st edition was checked out once, 3rd edition checked out 3x, interested in tracking 4 checkouts across all editions.  Will keep data granular to edition, but analysis will be collapsed on unique titles.  work_key has been generated in the 02_cleaning notebook to reflect this.

# Title Cleaning

### Assess discrepancies between how book/ ebook titles, authors vary. try to get them as close as possible for string matching between the two

In [12]:
#import data

book_checkouts = physical_checkouts[
    (physical_checkouts['materialtype'] == 'BOOK')]

ebook_checkouts = digital_checkouts[
    (digital_checkouts['materialtype'] == 'EBOOK')]

### check book title formats

In [13]:
#identify common separators and how often they're used for formatting book title metadata

import re
from collections import Counter

def inspect_separator(sep, n=20):
    escaped = re.escape(sep)
    total_distinct = book_checkouts['title'].nunique()
    matches = book_checkouts[book_checkouts['title'].str.contains(escaped, na=False, regex=False)]
    distinct_matches = matches['title'].nunique()
    pct = distinct_matches / total_distinct * 100
    
    print(f"=== Titles containing '{sep}' ===")
    print(f"Distinct titles: {distinct_matches:,} of {total_distinct:,} ({pct:.1f}%)\n")
    
    sample = matches['title'].drop_duplicates().sample(min(n, distinct_matches))
    for title in sample:
        parts = [p.strip() for p in title.split(sep)]
        for i, part in enumerate(parts):
            print(f"  Part {i+1}: {part[:80]}")
        print()

for sep in ['/', ':', ';', ',', '||']:
    inspect_separator(sep)
    print('=' * 60)

=== Titles containing '/' ===
Distinct titles: 469,746 of 517,224 (90.8%)

  Part 1: The art and craft of judging; the decisions of Judge Learned Hand
  Part 2: edited and annotated, with an introd., by Hershel Shanks.

  Part 1: How to change minds about our changing climate : let science do the talking the 
  Part 2: Seth B. Darling and Douglas L. Sisterson ; illustrations by Sarah M. Sisterson.

  Part 1: Great football writing : Sports illustrated 1954-2006
  Part 2: edited by Rob Fleder.

  Part 1: Inner way : Eastern Christian spiritual direction
  Part 2: Joseph J. Allen.

  Part 1: What is the AIDS crisis?
  Part 2: by Nico Medina ; illustrated by Tim Foley.

  Part 1: Pauline Kael : a life in the dark
  Part 2: Brian Kellow.

  Part 1: 1776
  Part 2: David McCullough.

  Part 1: Matisse
  Part 2: Diebenkorn
  Part 3: edited by Janet Bishop and Katherine Rothkopf.

  Part 1: The Paragon Hotel
  Part 2: Lyndsay Faye.

  Part 1: Some Spanish-American poets
  Part 2: translated by

In [14]:
# evaluate how often splitting title on / will be correct

# Split on / and evaluate each part
def evaluate_slash_splits(df, title_col='title'):
    titles_with_slash = df[df[title_col].str.contains('/', na=False, regex=False)][title_col].drop_duplicates()
    
    total = len(titles_with_slash)
    print(f"Total distinct titles with '/': {total:,}\n")
    
    # Heuristics for when / is likely IN the title vs a separator
    likely_in_title = titles_with_slash[
        titles_with_slash.str.contains(r'\d+/\d+', na=False)  # fractions like 1/2
        | titles_with_slash.str.contains(r'[A-Z]/[A-Z]', na=False)  # acronyms like AC/DC
        | titles_with_slash.str.contains(r'\w/\w', na=False)  # no spaces around slash
    ]
    
    likely_separator = titles_with_slash[
        titles_with_slash.str.contains(r'\s/\s', na=False)  # spaces on both sides
    ]

    print(f"Likely separator (spaces on both sides):  {len(likely_separator):,} ({len(likely_separator)/total*100:.1f}%)")
    print(f"Likely in title (no spaces / acronym):    {len(likely_in_title):,} ({len(likely_in_title)/total*100:.1f}%)")
    print(f"Ambiguous (other):                        {total - len(likely_separator) - len(likely_in_title):,} ({(total - len(likely_separator) - len(likely_in_title))/total*100:.1f}%)")

    print(f"\n=== Sample: likely separator ===")
    for title in likely_separator.sample(min(10, len(likely_separator))):
        parts = title.split('/')
        print(f"  Part 1: {parts[0].strip()[:60]}")
        print(f"  Part 2: {'/'.join(parts[1:]).strip()[:60]}")
        print()

    print(f"=== Sample: likely in title ===")
    for title in likely_in_title.sample(min(10, len(likely_in_title))):
        print(f"  {title[:80]}")
        print()

    #return ambiguous responses

    ambiguous = titles_with_slash[
        ~titles_with_slash.isin(likely_separator) & 
        ~titles_with_slash.isin(likely_in_title)
    ]
    
    print(f"=== Sample: ambiguous ===")
    for title in ambiguous.sample(min(10, len(ambiguous))):
        print(f"  {title[:80]}")
        print()
    
    return ambiguous

ambiguous = evaluate_slash_splits(book_checkouts)

Total distinct titles with '/': 469,746

Likely separator (spaces on both sides):  468,925 (99.8%)
Likely in title (no spaces / acronym):    2,610 (0.6%)
Ambiguous (other):                        -1,789 (-0.4%)

=== Sample: likely separator ===
  Part 1: Invisible men : the trailblazing Black artists of comic book
  Part 2: Ken Quattro.

  Part 1: Alfie's ABC
  Part 2: Shirley Hughes.

  Part 1: The nightmare brigade. Vol. 2, Into the woods
  Part 2: story, Franck Thilliez ; art, Yomgui Dumont ; color, Drac.

  Part 1: Tin man
  Part 2: Sarah Winman.

  Part 1: Arts and crafts rugs for craftsman interiors : the Crab Tree
  Part 2: Linda Parry and David Cathers ; with Diane Boucher, Ann Lane

  Part 1: Alice Sadie Celine
  Part 2: Sarah Blakley-Cartwright.

  Part 1: Ireland
  Part 2: author: Allanah Hopkin.

  Part 1: Jana goes wild
  Part 2: Farah Heron.

  Part 1: Shashin o tsukatte ehon o tsukurō !
  Part 2: Nakagawa Motoko, Wada Naoto.

  Part 1: Firestorm
  Part 2: Rachel Grant.



### only returning the first part of the book title before ' / ' separator 
### addresses 99.8% of its use as the most common separator character.  

### check ebook formats

In [15]:
#identify common separators and how often they're used for formatting book title metadata

import re
from collections import Counter

def inspect_separator(sep, n=20):
    escaped = re.escape(sep)
    total_distinct = ebook_checkouts['title'].nunique()
    matches = ebook_checkouts[ebook_checkouts['title'].str.contains(escaped, na=False, regex=False)]
    distinct_matches = matches['title'].nunique()
    pct = distinct_matches / total_distinct * 100
    
    print(f"=== Titles containing '{sep}' ===")
    print(f"Distinct titles: {distinct_matches:,} of {total_distinct:,} ({pct:.1f}%)\n")
    
    sample = matches['title'].drop_duplicates().sample(min(n, distinct_matches))
    for title in sample:
        parts = [p.strip() for p in title.split(sep)]
        for i, part in enumerate(parts):
            print(f"  Part {i+1}: {part[:80]}")
        print()

for sep in ['/', ':', ';', ',', '||']:
    inspect_separator(sep)
    print('=' * 60)

=== Titles containing '/' ===
Distinct titles: 1,310 of 270,972 (0.5%)

  Part 1: Dolphins
  Part 2: Los delfines

  Part 1: Making New Friends with Baby King
  Part 2: 和小金金做朋友

  Part 1: Tempted: Little Miss Innocent?
  Part 2: Annie, Get Your Guy
  Part 3: Messing Around with Max

  Part 1: Lolo and Birdie: I Want More!
  Part 2: ¡Quiero Más!

  Part 1: laget hiyt toxwum
  Part 2: Herring to Huckleberries

  Part 1: Home for the Holidays: The Forgetful Bride
  Part 2: When Christmas Comes

  Part 1: Denver
  Part 2: Boulder and Colorado's Front Range: A Comprehensive Hiking Guide

  Part 1: Butterflies on Carmen Street
  Part 2: Mariposas en la Calle Carmen

  Part 1: ¡Este es el día!
  Part 2: This Is the Day! (Bilingual)

  Part 1: Sizzle: Buchanan
  Part 2: Renard
  Part 3: MacKenna Series, Book 8

  Part 1: Phantom Waltz: Kendrick
  Part 2: Coulter
  Part 3: Harrigan Series, Book 2

  Part 1: Shielding Riley: Army Delta Force
  Part 2: Military Romance

  Part 1: 龟兔赛跑
  Part 2: T

### ebook titles don't seem to have separators with additional metadata and just contain the titles/ captions 

## EBook Title Cleaning - check frequency of parentheses containing noise

In [16]:
processed_dir = '/Users/audriswong/data-portfolio/projects/seattle-checkouts/data/processed'

#import data
book_checkouts = pd.read_parquet(f'{processed_dir}/book_checkouts.parquet').rename(columns={'title_clean': 'title'}) 
ebook_checkouts = pd.read_parquet(f'{processed_dir}/ebook_checkouts.parquet').rename(columns={'title_clean': 'title'})

In [17]:
##worth removing extra content in the parentheses?  e.g. (Pulitzer Prize, Oprah's Book Club)

import re
from collections import Counter

# Extract all parenthetical content from ebook titles
parens_content = []

for title in ebook_checkouts['title'].dropna():
    matches = re.findall(r'\(([^)]+)\)', str(title))
    parens_content.extend(matches)

parens_counts = Counter(parens_content)
print(f"=== Ebook Title Breakdown ===")
print(f"Total parenthetical instances: {len(parens_content):,}")
print(f"Distinct parenthetical values: {len(parens_counts):,}")
print("\n=== Most common parenthetical content ===")
for content, count in parens_counts.most_common(30):
    print(f"{count:>8,} | {content[:80]}")

#book titles
parens_content = []

for title in book_checkouts['title'].dropna():
    matches = re.findall(r'\(([^)]+)\)', str(title))
    parens_content.extend(matches)

parens_counts = Counter(parens_content)
print(f"=== Book Title Breakdown ===")
print(f"Total parenthetical instances: {len(parens_content):,}")
print(f"Distinct parenthetical values: {len(parens_counts):,}")
print("\n=== Most common parenthetical content ===")
for content, count in parens_counts.most_common(30):
    print(f"{count:>8,} | {content[:80]}")

=== Ebook Title Breakdown ===
Total parenthetical instances: 112,073
Distinct parenthetical values: 3,750

=== Most common parenthetical content ===
   2,668 | A GRAPHIC NOVEL
   1,636 | WORLD DIGITAL LIBRARY EDITION
   1,608 | AUBREY/MATURIN NOVELS
   1,573 | THOMAS & FRIENDS
   1,513 | 2003
   1,276 | BARBIE
     970 | ALMOST
     839 | 2014
     820 | MOSTLY
     747 | WITH BONUS CONTENT
     703 | 2012
     691 | PENGUIN CLASSICS DELUXE EDITION
     650 | STAR WARS
     646 | 2015
     588 | WITH AUDIO RECORDING
     575 | NOVEL
     571 | REVISED AND UPDATED
     565 | DC SUPER FRIENDS
     538 | WORLD DIGITAL LIBRARY
     489 | BOOK 1
     486 | PULITZER PRIZE WINNER
     472 | OPRAH'S BOOK CLUB
     470 | A NOVEL
     443 | 2013
     433 | SESAME STREET
     407 | MOVIE TIE-IN
     404 | SPANISH EDITION
     384 | READ-ALOUD
     383 | SCHOLASTIC GOLD
     365 | A COOKBOOK
=== Book Title Breakdown ===
Total parenthetical instances: 73,089
Distinct parenthetical values: 3,298

==

In [18]:
# Classify parenthetical content into categories
def classify_paren(content):
    c = content.lower().strip()
    
    if re.search(r'\d{4}', c):
        return 'Year/Date'
    elif any(w in c for w in ['prize', 'award', 'winner', 'pulitzer', 'booker', 'newbery', 'caldecott', 'national book']):
        return 'Award/Prize'
    elif any(w in c for w in ['book club', 'reese', 'oprah', 'good morning america', 'gma', 'today show']):
        return 'Book Club'
    elif any(w in c for w in ['audio', 'unabridged', 'abridged', 'narrated', 'read by']):
        return 'Audio Details'
    elif any(w in c for w in ['bonus', 'excerpt', 'preview', 'includes', 'with ', 'enhanced', 'extra']):
        return 'Bonus Content'
    elif any(w in c for w in ['book 1', 'book 2', 'book 3', 'vol', 'volume', 'part ', 'series', '#']):
        return 'Series/Volume'
    elif any(w in c for w in ['edition', 'revised', 'updated', 'expanded', 'anniversary']):
        return 'Edition Info'
    elif any(w in c for w in ['translation', 'translated', 'adapted']):
        return 'Translation'
    elif any(w in c for w in ['illustrated', 'illustrations', 'pictures']):
        return 'Illustration Info'
    elif any(w in c for w in ['novel', 'memoir', 'story', 'tales', 'poems', 'essays']):
        return 'Likely Title Content'
    elif len(c.split()) <= 3:
        return 'Short — likely title content'
    else:
        return 'Other/Unclear'

# Apply classification
paren_classified = Counter()
paren_examples   = {}

for title in ebook_checkouts['title'].dropna():
    matches = re.findall(r'\(([^)]+)\)', str(title))
    for match in matches:
        category = classify_paren(match)
        paren_classified[category] += 1
        if category not in paren_examples:
            paren_examples[category] = []
        if len(paren_examples[category]) < 5:
            paren_examples[category].append(f"{match[:60]} <- {title[:60]}")

total = sum(paren_classified.values())
print(f"=== Parenthetical Content Categories ===\n")
for category, count in sorted(paren_classified.items(), key=lambda x: x[1], reverse=True):
    pct = count / total * 100
    print(f"{count:>8,} ({pct:>5.1f}%)  {category}")
    for ex in paren_examples.get(category, []):
        print(f"            e.g. {ex}")
    print()

=== Parenthetical Content Categories ===

  42,457 ( 37.9%)  Short — likely title content
            e.g. AND JOKES! <- HEIDEGGER AND A HIPPO WALK THROUGH THOSE PEARLY GATES: USING
            e.g. STILL <- HOW TO BE EVERYTHING: A GUIDE FOR THOSE WHO (STILL) DON'T KN
            e.g. ALMOST <- THE JESUIT GUIDE TO (ALMOST) EVERYTHING: A SPIRITUALITY FOR 
            e.g. STILL <- YES WE (STILL) CAN: POLITICS IN THE AGE OF OBAMA, TWITTER, A
            e.g. FAIRLY <- RED: THE (FAIRLY) TRUE TALE OF RED RIDING HOOD

  23,087 ( 20.6%)  Other/Unclear
            e.g. DC SUPER HERO GIRLS <- HARLEY QUINN AT SUPER HERO HIGH (DC SUPER HERO GIRLS)
            e.g. NOT NECESSARILY IN THAT ORDER <- TRADING BASES: A STORY ABOUT WALL STREET, GAMBLING, AND BASE
            e.g. THE FAULT IN OUR STARS <- BAJO LA MISMA ESTRELLA: (THE FAULT IN OUR STARS)
            e.g. BUT NOT TOO GOOD-FOR-YOU <- HEALTHYISH: A COOKBOOK WITH SERIOUSLY SATISFYING, TRULY SIMP
            e.g. AND OTHER GOOD STUFF <- EXCE

### not worth removing all content within ().  ~80% of parentheses content within ebook titles are actually parts of the title